### PR TITLE
Raise error for duplicate headers in hash mode (fixes #44)

### LINKED ISSUE #44: Wrongly parsed hash
The hash parsing the [F2022-10.xlsx](https://github.com/martijn/xsv/files/9700976/F2022-10.xlsx) is wrong:

~~~
$ pry
[1] pry(main)> require 'xsv'
=> true
[2] pry(main)> x = Xsv.open 'F2022-10.xlsx', :parse_headers => true
=> #<Xsv::Workbook:300 sheets=1 trim_empty_rows=false>
[3] pry(main)> s = x.sheets.first
=> #<Xsv::Sheet:320 mode=hash>
[4] pry(main)> s[0]
=> {"Interní číslo"=>"2022009",
 nil=>61,
 "Typ faktury"=>"FAKTURA",
 "Název firmy nebo jméno osoby"=>"Foo bar",
 "IČO"=>1234567,
 "DIČ"=>"CZ1234567",
 "Ulice"=>nil,
 "PSČ"=>"Velehradská 1",
 "Město"=>"686 03",
 "Datum vystavení"=>"Staré Město",
 "Datum splatnosti"=>nil,
 "Měna"=>#<Date: 2022-08-01 ((2459793j,0s,0n),+0s,2299161j)>,
 "Pořadí"=>#<Date: 2022-08-15 ((2459807j,0s,0n),+0s,2299161j)>,
 "Název"=>nil,
 "Množství"=>"EUR",
 "Cena za MJ"=>nil}
~~~

The correct output should be:

~~~diff
--- wrong
+++ correct
@@ -7,18 +7,17 @@
 => #<Xsv::Sheet:320 mode=hash>
 [4] pry(main)> s[0]
 => {"Interní číslo"=>"2022009",
- nil=>61,
  "Typ faktury"=>"FAKTURA",
  "Název firmy nebo jméno osoby"=>"Foo bar",
  "IČO"=>1234567,
  "DIČ"=>"CZ1234567",
- "Ulice"=>nil,
- "PSČ"=>"Velehradská 1",
- "Město"=>"686 03",
- "Datum vystavení"=>"Staré Město",
- "Datum splatnosti"=>nil,
- "Měna"=>#<Date: 2022-08-01 ((2459793j,0s,0n),+0s,2299161j)>,
- "Pořadí"=>#<Date: 2022-08-15 ((2459807j,0s,0n),+0s,2299161j)>,
- "Název"=>nil,
- "Množství"=>"EUR",
- "Cena za MJ"=>nil}
+ "Ulice"=>"Velehradská 1",
+ "PSČ"=>"686 03",
+ "Město"=>"Staré Město",
+ "Datum vystavení"=>#<Date: 2022-08-01 ((2459793j,0s,0n),+0s,2299161j)>,
+ "Datum splatnosti"=>#<Date: 2022-08-15 ((2459807j,0s,0n),+0s,2299161j)>,
+ "Měna"=>"EUR",
+ "Pořadí"=>1,
+ "Název"=>"Item 1,
+ "Množství"=>2,
+ "Cena za MJ"=>61}
~~~

I think that the parser is confused by the empty columns.

### PR DESCRIPTION
**Pull request by martijn**
As discussed in #44 the current behaviour for hash mode has the potential leads to confusion.

This change makes Xsv raise an error when trying to enter hash mode on a sheet with duplicate non-nil header values. I left nil out of scope because it's common for Excel sheets to have empty columns for spacing or trailing on the end of the sheet.

I'm not sure if we really want this behaviour, and I'm planning on bumping the minor version to signal the potential breakage this can cause. @shkm @voxik looking forward to your comments on this one.

### CODE REVIEW

**Comment by voxik** [comment]
Although it does not address the #44, I like this :+1: I think it is better to fail then possibly deliver corrupted data.

File: lib/xsv/sheet.rb [hunk 1]
```diff
@@ -83,6 +83,13 @@ def [](number)
     # @return [self]
     def parse_headers!
       @headers = parse_headers
+
+      # Check for duplicate headers, but don't care about nil columns
+      if (unique_headers = @headers.compact.uniq) != @headers.compact
+        first_dupe = @headers.compact.zip(unique_headers).detect { |k,v| k != v }.first
+        raise Xsv::DuplicateHeaders, "Duplicate header '#{first_dupe}' found, consider parsing this sheet in array mode."
+      end
```
Comment (shkm) [inline lib/xsv/sheet.rb:91]: We could simplify this quite a bit, I think.

```
first_duplicate = @headers.detect { |header| @headers.count(header) > 1 }

if (first_duplicate)
  raise Xsv::DuplicateHeaders, "Duplicate header '#{first_duplicate}' found, consider parsing this sheet in array mode."
end
```

**Review by shkm — CHANGES_REQUESTED** [review]

**Comment by shkm** [comment]
+1 but I think it's worth making that change so we don't have to look up  zip again ;-).

**Comment by martijn** [comment]
> +1 but I think it's worth making that change so we don't have to look up zip again ;-).

Ohh yeah that is much better, thanks!

File: lib/xsv/sheet.rb (hunk 1 above)
Comment (martijn) [reply lib/xsv/sheet.rb:91]: Thanks, much better

**Pull request merged**